### PR TITLE
Abstract class for target/aux computation

### LINKED ISSUE #1179: Abstract class for the teacher in student-teacher training
General class that can run computation for target computation and auxiliary stuff

### PR DESCRIPTION
**Pull request by sophie-xhonneux**
Implemented Identity class

## Description

Adds a Target class with an identity to prepare for student-teacher training

## Issue Number

Closes #1179 

Is this PR a draft? Mark it as draft.

## Tests run 

- [x] launch-slurm.py
- [x] train forecasting
- [x] train_continue
- [x] inference (via the integration-test)
- [ ] backwards compatibility [NO]

## Checklist before asking for review

-   [x] I have performed a self-review of my code
-   [x] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [x] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [x] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

### CODE REVIEW

File: src/weathergen/train/target_and_aux_ssl_teacher.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
+from typing import Any
+
+from weathergen.train.target_and_aux_module_base import TargetAndAuxModuleBase
+
+
+class EMATeacher(TargetAndAuxModuleBase):
+    def __init__(self, model, rng, ema_model, batch_size, **kwargs):
+        # One of the issues is that the teacher model may have a different architecture
```
Comment (shmh40) [inline src/weathergen/train/target_and_aux_ssl_teacher.py]: Do you mean that e.g. in JEPA the student has the predictor too?

Comment (sophie-xhonneux) [reply src/weathergen/train/target_and_aux_ssl_teacher.py]: Yea, in JEPA the student is  Predictor(Encoder(x')) whereas the teacher is just Encoder(x), but also in BYOL there is a difference for instance

Comment (shmh40) [reply src/weathergen/train/target_and_aux_ssl_teacher.py]: Cool. Is there a useful abstraction we could stick with that would be helpful -- always EMA'ed encoder for example? EMATeacherEncoder always the same, then add e.g. predictor to this? This might not help, and don't know if this holds for byol, just thinking

File: src/weathergen/model/model.py [hunk 2]
```diff
@@ -907,3 +909,17 @@ def predict(
             preds_tokens += [checkpoint(self.pred_heads[ii], tc_tokens, use_reentrant=False)]
 
         return preds_tokens
+
+
+def get_model(student_or_teacher, cf: Config, sources_size, targets_num_channels, targets_coords_size, **kwargs):
```
Comment (clessig) [inline src/weathergen/model/model.py]: `instantiate_model()` is a more natural name for me

File: src/weathergen/model/ema.py [hunk 3]
```diff
@@ -44,7 +44,7 @@ def reset(self):
         self.ema_model.to_empty(device="cuda")
         maybe_sharded_sd = self.original_model.state_dict()
         # this copies correctly tested in pdb
-        mkeys, ukeys = self.ema_model.load_state_dict(maybe_sharded_sd, strict=True, assign=False)
+        mkeys, ukeys = self.ema_model.load_state_dict(maybe_sharded_sd, strict=False, assign=False)
```
Comment (clessig) [inline src/weathergen/model/ema.py:47]: Why is this changed?

File: src/weathergen/model/model.py [hunk 4]
```diff
@@ -907,3 +909,17 @@ def predict(
...
+    if student_or_teacher == "student" or student_or_teacher == "teacher":
+        return Model(cf, sources_size, targets_num_channels, targets_coords_size).create()
+    else:
+        if cf["training_mode"] == "masking": # TODO implement mode "student-teacher-pretrain":
```
Comment (clessig) [inline src/weathergen/model/model.py]: This should be a nested dict. But we should write an example config to see how it looks and feels like and how it works.

File: src/weathergen/train/target_and_aux_module_base.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+    def update_state_post_opt_step(self, istep, batch, model, **kwargs) -> None:
+        pass
+
+    def compute(self, *args, **kwargs) -> tuple[Any, Any]:
+        pass
+
+
+class IdentityTargetAndAux(TargetAndAuxModuleBase):
+    def __init__(self, model, rng, config):
```
Comment (clessig) [inline src/weathergen/train/target_and_aux_module_base.py]: Could we have a brief documentation

File: src/weathergen/train/target_and_aux_ssl_teacher.py (hunk 1 above)
Comment (clessig) [reply src/weathergen/train/target_and_aux_ssl_teacher.py]: I agree. The predictor could be the identity if it's not present.

File: src/weathergen/train/target_and_aux_ssl_teacher.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+        # to the student, e.g. JEPA. So we need quite a flexible way to instantiate the 
+        # the teacher. Because of the device sharding etc that requires quite a bit of
+        # massaging we assume that the teacher creates the EMA model correctly. However,
+        # note that you cannot assume that model.state_dict equals ema_model.state_dict
```
Comment (clessig) [inline src/weathergen/train/target_and_aux_ssl_teacher.py]: I understand why student and teacher can have different state_dicts. But why should student/teacher and ema_student/teacher differ?

File: src/weathergen/model/model.py (hunk 2 above)
Comment (clessig) [reply src/weathergen/model/model.py]: And I don't think it should go to `model.py`. If we have the function then it seems more natural that it is also responsible which model potentially to instantiate.

File: src/weathergen/train/trainer.py [hunk 7]
```diff
@@ -157,13 +157,15 @@ def inference(self, cf, devices, run_id_trained, epoch):
         self.validate(epoch=0)
         logger.info(f"Finished inference run with id: {cf.run_id}")
 
-    def init_model_and_shard(self, cf, devices):
+    def init_model_and_shard(self, cf, student_or_teacher, devices):
```
Comment (clessig) [inline src/weathergen/train/trainer.py]: Can we take this function out of the trainer class. It's for me closer to the model infra than to orchestrating the run (which is the purpose of the Trainer). I would put it in the same file as `instantiate_model()`

File: src/weathergen/train/trainer.py [hunk 8]
```diff
@@ -313,8 +315,20 @@ def run(self, cf, devices, run_id_contd=None, epoch_contd=None):
 
         self.validate_with_ema = cf.get("validate_with_ema", False)
         self.ema_model = None
+        # validate_with_ema is incompatible with student-teacher
+        self.validate_with_ema = False  # TODO remove for testing only
```
Comment (clessig) [inline src/weathergen/train/trainer.py]: Let's not forget it. Can't you just disable it in your test config?

File: src/weathergen/train/trainer.py [hunk 9]
```diff
@@ -313,8 +315,20 @@ def run(self, cf, devices, run_id_contd=None, epoch_contd=None):
...
-            meta_ema_model = self.init_model_and_shard(cf, devices)[0]
+            meta_ema_model = self.init_model_and_shard(cf, "student", devices)[0]
+            self.ema_model = EMAModel(
+                self.model,
+                meta_ema_model,
+                halflife_steps=cf.get("ema_halflife_in_thousands", 1e-3),
+                rampup_ratio=cf.get("ema_ramp_up_ratio", 0.09),
+                is_model_sharded=(cf.with_ddp and cf.with_fsdp),
+            )
+        elif cf["training_mode"] == "masking":  # "student-teacher-pretrain":
```
Comment (clessig) [inline src/weathergen/train/trainer.py]: This is currently a proxy? Because an EMATeacher doesn't make sense for masking and it should be the identity then, not?

File: src/weathergen/train/trainer.py [hunk 10]
```diff
@@ -591,14 +613,20 @@ def train(self, epoch):
...
                     self.model_params, batch, cf.forecast_offset, forecast_steps
                 )
+
+                targets, aux_outputs = self.target_and_aux_calculator.compute(
+                    bidx, batch, self.model_params, self.model, cf.forecast_offset, forecast_steps
+                )
             loss_values = self.loss_calculator.compute_loss(
                 preds=preds,
-                streams_data=batch[0],
+                streams_data=batch[0],  # should additionally take targets?
```
Comment (clessig) [inline src/weathergen/train/trainer.py]: Yes, this should take targets. We should have an `TargetAndAuxCalculatorIdentity` class that takes the batch and returns just the physical space targets. (No strong feelings if we call TargetAndAuxCalculatorIdentity or TargetAndAuxCalculatorPhysical or something similar)

File: src/weathergen/train/trainer.py [hunk 11]
```diff
@@ -591,14 +613,20 @@ def train(self, epoch):
...
             )
             if cf.latent_noise_kl_weight > 0.0:
                 kl = torch.cat([posterior.kl() for posterior in posteriors])
                 loss_values.loss += cf.latent_noise_kl_weight * kl.mean()
 
+            self.target_and_aux_calculator.update_state_pre_backward(bidx, batch, self.model)
```
Comment (clessig) [inline src/weathergen/train/trainer.py:496]: The dependence should be on the optimizer step but not on the bidx, not?

File: src/weathergen/train/trainer.py [hunk 12]
```diff
@@ -622,14 +650,16 @@ def train(self, epoch):
             self.grad_scaler.update()
             # self.optimizer.step()
 
+            self.target_and_aux_calculator.update_state_post_opt_step(bidx, batch, self.model)
```
Comment (clessig) [inline src/weathergen/train/trainer.py:521]: Same as above: why is the dependence on bidx and not cf.istep?

File: src/weathergen/train/trainer.py [hunk 13]
```diff
@@ -622,14 +650,16 @@ def train(self, epoch):
...
+
             # update learning rate
             self.lr_scheduler.step()
 
             # EMA update
             if self.validate_with_ema:
                 self.ema_model.update(
-                    self.cf.istep * self.world_size_original * self.cf.batch_size_per_gpu,
-                    self.world_size_original * self.cf.batch_size_per_gpu,
+                    self.cf.istep * get_batch_size(self.cf, self.world_size_original),
```
Comment (clessig) [inline src/weathergen/train/trainer.py:529]: We need to abstract this into a function in `utils/distributed.py`

File: src/weathergen/train/trainer_base.py [hunk 14]
```diff
@@ -167,3 +170,14 @@ def get_perf(self):
             perf_mem /= len(self.device_handles)
 
         return perf_gpu, perf_mem
+
+
+# should be moved to its own file so as to prevent cyclical imports
+def get_target_and_aux_calculator(config, model, rng, batch_size, **kwargs):
```
Comment (clessig) [inline src/weathergen/train/trainer_base.py]: This should go to the same file as `instantiate_model.py`.

**Review by clessig — COMMENTED** [review]
Looks already very nice overall but some minor structural changes would be good, see detailed comments.

File: src/weathergen/train/target_and_aux_ssl_teacher.py (hunk 1 above)
Comment (sophie-xhonneux) [reply src/weathergen/train/target_and_aux_ssl_teacher.py]: We will need different "heads" for different latent student-teacher losses, the predictor would be just one of them

File: src/weathergen/model/model.py (hunk 2 above)
Comment (sophie-xhonneux) [reply src/weathergen/model/model.py]: it felt unnecessary to create another file for it

File: src/weathergen/model/ema.py (hunk 3 above)
Comment (sophie-xhonneux) [reply src/weathergen/model/ema.py:47]: because teacher arch =/= student arch so it cannot be strict

File: src/weathergen/train/target_and_aux_ssl_teacher.py (hunk 6 above)
Comment (sophie-xhonneux) [reply src/weathergen/train/target_and_aux_ssl_teacher.py]: can you elaborate? I don't understand your comment, the teacher is a model and the student is a model, it just so happens that the teacher model is an EMA

File: src/weathergen/train/trainer.py (hunk 7 above)
Comment (sophie-xhonneux) [reply src/weathergen/train/trainer.py]: it does need to be piped through, we can hide it in config if you prefer

File: src/weathergen/train/trainer.py (hunk 8 above)
Comment (sophie-xhonneux) [reply src/weathergen/train/trainer.py]: easy to search for all TODOs before merging :)

File: src/weathergen/train/trainer.py (hunk 9 above)
Comment (sophie-xhonneux) [reply src/weathergen/train/trainer.py]: yes just a proxy for testing, before I write all the boring config stuff

File: src/weathergen/train/trainer.py (hunk 11 above)
Comment (sophie-xhonneux) [reply src/weathergen/train/trainer.py:496]: yes

File: src/weathergen/train/trainer.py (hunk 13 above)
Comment (sophie-xhonneux) [reply src/weathergen/train/trainer.py:529]: this change does this abstraction, not sure I understand

File: src/weathergen/train/trainer_base.py (hunk 14 above)
Comment (sophie-xhonneux) [reply src/weathergen/train/trainer_base.py]: sure, how strongly are you married to `instantiate_model`?

File: src/weathergen/train/loss_modules/loss_module_latent.py [hunk 15]
```diff
@@ -0,0 +1,112 @@
+# ruff: noqa: T201
```
Comment (clessig) [inline src/weathergen/train/loss_modules/loss_module_latent.py]: Let's remove this. This will come with the diffusion model when it's actually needed (and working).

File: src/weathergen/train/loss_modules/loss_module_ssl.py [hunk 16]
```diff
@@ -0,0 +1,38 @@
+# ruff: noqa: T201
```
Comment (clessig) [inline src/weathergen/train/loss_modules/loss_module_ssl.py]: Same as above. Let's merge when it's needed/will be used.

File: src/weathergen/model/model.py [hunk 17]
```diff
@@ -875,3 +876,26 @@ def predict(
             preds_tokens += [checkpoint(self.pred_heads[ii], tc_tokens, use_reentrant=False)]
 
         return preds_tokens
+
+
+def get_model(
+    student_or_teacher,
+    cf: Config,
+    sources_size,
```
Comment (clessig) [inline src/weathergen/model/model.py]: Reminder for the future: It's not very nice how this is handled at the moment and passed around.

File: src/weathergen/model/model.py [hunk 18]
```diff
@@ -875,3 +876,26 @@ def predict(
             preds_tokens += [checkpoint(self.pred_heads[ii], tc_tokens, use_reentrant=False)]
 
         return preds_tokens
+
+
+def get_model(
+    student_or_teacher,
```
Comment (clessig) [inline src/weathergen/model/model.py]: I am struggling a bit if this is a good interface. Should we directly ask for model, encoder here? What we want is:
- student teacher
  - The TargetAuxCalculator calls get_model() there? 
  - What about the model-model? In Trainer?
- Diffusion
  - Similar to the above
- Masked token modeling
  - TargetAuxCalculator is the identity

File: src/weathergen/train/target_and_aux_module_base.py [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+        pass
+
+    def compute(self, *args, **kwargs) -> tuple[Any, Any]:
+        pass
+
+    def to_device(self, device):
+        pass
+
+
+class IdentityTargetAndAux(TargetAndAuxModuleBase):
```
Comment (clessig) [inline src/weathergen/train/target_and_aux_module_base.py]: We should rename this `PhysicalTargetAndAux` or something similar. It's conceptual the identity but it will extract the data from the StreamData object.

File: src/weathergen/train/loss_modules/loss_module_latent.py (hunk 15 above)
Comment (sophie-xhonneux) [reply src/weathergen/train/loss_modules/loss_module_latent.py]: Not sure I understand

File: src/weathergen/train/target_and_aux_module_base.py (hunk 19 above)
Comment (sophie-xhonneux) [reply src/weathergen/train/target_and_aux_module_base.py]: changed

File: src/weathergen/model/model.py (hunk 18 above)
Comment (sophie-xhonneux) [reply src/weathergen/model/model.py]: Maybe let's chat on Monday. I didn't have a particular striking idea for how to do this best
